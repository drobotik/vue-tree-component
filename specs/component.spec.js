import {mount} from '@vue/test-utils';
import component from '@/Node.vue'
import data from './data.json';

describe( "component", () => {
    const props = {
        properties: { expanded: false },
        attributes: { name: 'the name'},
        children: [ {properties: { expanded: false } } ]
    }

    it("default properties", () => {
        const node = mount(component)
        expect(node.props()).toMatchSnapshot()
    })

    it('expand arrow not showing if empty children', async () => {
        const node = mount(component, {props: { children: [] }})
        expect(node.find('.arrow').exists()).toBeFalsy()
        await node.setProps({ children : [{}]})
        expect(node.find('.arrow').exists()).toBeTruthy()
    })

    it('expand classes', async () => {
        const $store = { dispatch: jest.fn()}
        const node = mount(component, {global: {mocks: { $store }}})
        expect(node.find('.children').exists()).toBeFalsy()
        expect(node.find('.arrow').exists()).toBeFalsy()
        await node.setProps({ properties: { expanded: true }, attributes: { name: 'test'}, children: [{}] })
        expect(node.find('.children').exists()).toBeTruthy()
        expect(node.find('.arrow').classes('arrow-right')).toBeFalsy()
        expect(node.find('.arrow').classes('arrow-down')).toBeTruthy()
        await node.setProps({ properties: { expanded: false }, attributes: { name: 'test'}, children: [{}] })
        expect(node.find('.children').exists()).toBeFalsy()
        expect(node.find('.arrow').classes('arrow-right')).toBeTruthy()
        expect(node.find('.arrow').classes('arrow-down')).toBeFalsy()
    })

    it('expand event', async() => {
        const $store = { dispatch: jest.fn()}
        const mockMethod = jest.spyOn(component.methods, 'expand')
        await mount(component,{global: { mocks: { $store } }, props}).find('.arrow').trigger('click')
        expect(mockMethod).toHaveBeenCalledTimes(1);
    })

    it('expand method', async() => {
        const $store = { dispatch: jest.fn() }
        const node = mount(component, {global: { mocks: { $store } }, props})
        await node.find('.arrow').trigger('click')
        expect($store.dispatch).toHaveBeenCalledTimes(1);
        expect($store.dispatch).toHaveBeenCalledWith('expand', 'the name')
    })

    it("render data", () => {
        let tree = data[0];
        const node = mount(component, {
            props: {
                properties: tree.properties,
                attributes: tree.attributes,
                children: tree.children
            }
        })
        expect(node.html()).toMatchSnapshot()
    })
})