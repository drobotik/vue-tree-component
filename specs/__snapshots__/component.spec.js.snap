// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component default properties 1`] = `
Object {
  "attributes": Object {},
  "children": Array [],
  "depth": 0,
  "properties": Object {
    "expanded": false,
    "selected": false,
  },
}
`;

exports[`component render data 1`] = `
<div>
  <div><span class="arrow arrow-down"></span><span>Books</span></div>
  <div class="children">
    <div>
      <div>
        <!--v-if--><span>Fantastic</span>
      </div>
      <!--v-if-->
    </div>
    <div>
      <div>
        <!--v-if--><span>Fantasy</span>
      </div>
      <!--v-if-->
    </div>
    <div>
      <div><span class="arrow arrow-right"></span><span>Adventure</span></div>
      <!--v-if-->
    </div>
  </div>
</div>
`;
